test/e2e/azure_timesync.go

```diff
@@ -0,0 +1,94 @@
+// +build e2e
+
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package e2e
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	. "github.com/onsi/gomega"
+	corev1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/cluster-api/test/framework"
+	"sigs.k8s.io/cluster-api/util"
+	kinderrors "sigs.k8s.io/kind/pkg/errors"
+)
+
+// AzureTimeSyncSpecInput is the input for AzureTimeSyncSpec.
+type AzureTimeSyncSpecInput struct {
+	BootstrapClusterProxy framework.ClusterProxy
+	Namespace             *corev1.Namespace
+	ClusterName           string
+}
+
+// AzureTimeSyncSpec implements a test that verifies time synchronization is healthy for
+// the nodes in a cluster.
+func AzureTimeSyncSpec(ctx context.Context, inputGetter func() AzureTimeSyncSpecInput) {
+	var (
+		specName = "azure-timesync"
+		input    AzureTimeSyncSpecInput
+	)
+
+	input = inputGetter()
+	Expect(input.BootstrapClusterProxy).NotTo(BeNil(), "Invalid argument. input.BootstrapClusterProxy can't be nil when calling %s spec", specName)
+
+	namespace, name := input.Namespace.Name, input.ClusterName
+	managementClusterClient := input.BootstrapClusterProxy.GetClient()
+	machines, err := getMachinesInCluster(ctx, managementClusterClient, namespace, name)
+	Expect(err).NotTo(HaveOccurred(), "Failed to get machines for the %s/%s cluster", namespace, name)
+
+	testfuncs := []func() error{}
+	for i := range machines.Items {
+		m := &machines.Items[i]
+		Byf("checking that time synchronization is healthy on %s", m.Name)
+		cluster, err := util.GetClusterFromMetadata(ctx, managementClusterClient, m.ObjectMeta)
+		Expect(err).NotTo(HaveOccurred())
+		controlPlaneEndpoint := cluster.Spec.ControlPlaneEndpoint.Host
+		hostname := m.Spec.InfrastructureRef.Name
+		port := e2eConfig.GetVariable(VMSSHPort) // Or VMSSSSHPort when MachinePools are supported
+
+		execToStringFn := func(expected, command string, args ...string) func() error {
+			// don't assert in this test func, just return errors
+			return func() error {
+				f := &strings.Builder{}
+				if err := execOnHost(controlPlaneEndpoint, hostname, port, f, command, args...); err != nil {
+					return err
+				}
+				if !strings.Contains(f.String(), expected) {
+					return fmt.Errorf("expected \"%s\" in command output:\n%s", expected, f.String())
+				}
+				return nil
+			}
+		}
+
+		testfuncs = append(testfuncs,
+			execToStringFn(
+				"✓ chronyd is active",
```

Comment:
is the `✓` required / reliable? wondering if it might lead to weird behavior

Reply:
It's just a unicode check mark character. I put it in the "success" text to make it obviously distinct from any text that `systemctl is-active` might print.